proposed/local-sdk-global-json.md

```diff
@@ -0,0 +1,112 @@
+- [VS code](https://github.com/dotnet/vscode-csharp/issues/6471) having to adjust to consider local directories for SDK because our resolver can't find them.
+
+## Detailed Design
+The global.json file will support two new properties under the `sdk` object:
+
+- `"additionalPaths"`: this is a list of paths that the host resolver should consider when looking for compatible SDKs. Relative paths will be interpreted relative to the global.json file.
+- `"additionalPathsOnly"`: when `true` the resolver will _only_ consider paths in `additionalPaths`. It will not consider any machine wide locations (unless they are specified in `additionalPaths`). The default for this property is `false`.
+
+The `additionalPaths` works similar to how multi-level lookup works. It adds additional locations that the host resolver should consider when trying to resolve a compatible .NET SDK. For example:
+
+```json
+{
+    "sdk": {
+        "additionalPaths": [ ".dotnet" ],
+    }
+}
+```
+
+In this configuration the host resolver would find a compatible SDK, if it exists in `.dotnet` or a machine wide location. The host resolver will consider the `additionalPaths` in the order they are defined and will stop at the first match. If none of the locations have a matching SDK then
+ will fall back to the existing SDK resolution strategy: `%PATH%` followed by machine wide installations.
+
+This lookup will stop on the first match which means it won't necessarily find the best match. Consider a scenario with a global.json that has:
+
+```json
+{
+  "sdk": {
+    "additionalPaths": [ ".dotnet" ],
+    "version": "7.0.200",
+    "rollForward": "latestFeature"
+  }
+}
+```
+
+In a scenario where the `.dotnet` directory had 7.0.200 SDK but there was a machine wide install of 7.0.300 SDK, the host resolver would pick 7.0.200 out of `.dotnet`. That location is considered first, it has a matching SDK and hence discovery stops there.
+
+The values in the `additionalPaths` property can be a relative or absolute path. When a relative path is used it will be resolved relative to the location of global.json. These values also support the following substitutions:
+
+- `"$user$"`: this matches the user-local installation point of .NET SDK for the current operating system: `%LocalAppData%\Microsoft\dotnet` on Windows and `$HOME/.dotnet` on Linux/macOS.
+- `"$machine$"`: this matches the machine installation point of .NET for the current operating system.
+- `%VARIABLE%/$VARIABLE`: environment variables will be substituted. Either the Windows or Unix format can be used here and will be normalized for the operating system the host resolver executes on.
```

Comment:
What happens if the environment variable contains multiple paths (such as PATH)?

Would it support the `${VARIABLE}` syntax as well?

What happens if the path / environment variable doesn't exist? (One scenario would be if someone wanted to use an environment variable that's different between platforms - HOME vs USERPROFILE for instance).

Would either path separator (`\`, `/`) work on all platforms?

Reply:
Overall I feel that supporting env. variable expansion is unnecessary. What would be the typical use case for it?
I also agree with basically everything Rolf points out - especially the fact that the env. variables are unlikely to match between platforms, and as such it will be effectively impossible to use them for anything in a repo, only useful for local dev work where the dev authors `global.json` themselves. Not sure how common that would be, and it's a case where asking the developer to use other env. variables (`DOTNET_ROOT`, `PATH`) doesn't seem too bad.

Reply:
Agreed. I generally think the design already generally feels like it's encoding local dev machine setup into the global.json. While not ideal, it's a decent compromise since a lot of "common" practices would put an SDK locally into the repo layout at predictable locations. Adding environment variable expansion on top of that seems unnecessary. That said, why not add the flexibility? What's the harm?

Reply:
The harm is complexity of implementation - in a component which is not versioned in the normal sense (we always use the latest version on the machine), thus requiring effectively 100% backward compatibility all the time (or close to it). Additionally, this would require some additional string parsing logic (like ability escape the special characters, how to deal with errors and so on). Basically, introducing another parser, which is always a concern. Personally I think that if we can make this useful without env. var support, I think we should do that.

There's also the potential unintended side-effects:
* "My SDK doesn't work in one window, but works in the other"
* "What's your global.json"
* <file with env. vars>
* Oh, that's why :-(
Basically reintroducing parts of the problem which we have today with PATH and DOTNET_ROOT.

Reply:
That's a good argument. Leaning towards pulling env var support out.

Reply:
I'd also rather not add environment variable expansion until there's a more compelling scenario.

Reply:
Going to remove environment variables from the doc. Will add it into a _possible future expansions_ section.